python/pyarrow/tests/test_convert_builtin.py

```diff
@@ -2457,14 +2457,29 @@ def test_array_from_pylist_offset_overflow():
     )
 ])
 def test_array_accepts_pyarrow_scalar(seq, data, scalar_data, value_type):
-    if type(seq(scalar_data)) == set:
-        pytest.skip("TODO: The elements in the set get reordered.")
     expect = pa.array(data, type=value_type)
     result = pa.array(seq(scalar_data))
-    assert expect.equals(result)
+
+    if type(seq(scalar_data)) == set:
+        try:
+            import pyarrow.compute as pc
+            expect = [expect[a] for a in pc.array_sort_indices(expect).to_pylist()]
+            result = [result[a] for a in pc.array_sort_indices(result).to_pylist()]
+            expect == result
+        except pa.lib.ArrowNotImplementedError:
+            set(expect) == set(result)
```

Comment:
I think it's probably good enough to just always use the fallback (or could also be `sorted(expect.to_pylist()) == sorted(result.to_pylist())` ?)

(that would avoid some of the complexity)

Reply:
That was my first try, but in the case of dictionary type the set mixed the elements differently 🤷‍♀️ 
But `sorted(expect.to_pylist()) == sorted(result.to_pylist())` is a great idea! Tried without the `to_pylist` but it ofcourse didn't work. Will update today 👍

Reply:
There is always a case that doesn't fit in this PR 🤣  

```python
TypeError: unhashable type: 'list'
TypeError: unhashable type: 'dict'
```

I can do `try-except` and take the items of the lists in the case of a `TypeError` from above (in case of ListType and StructType as `to_pylist()` returns a dictionary in a list).

Reply:
> There is always a case that doesn't fit in this PR

Which case exactly gives that error message with ``sorted(expect.to_pylist()) == sorted(result.to_pylist())``?

Reply:
`ListType` and `StructType`, that is:
- `data = [(1, 'bar')], scalar_data = [<pyarrow.StructScalar: [('a', 1), ('b', 'bar')]>], value_type = StructType(struct<a: int8, b: string>)`
- `data = [(1, 2)], scalar_data = [<pyarrow.StructScalar: [('a', 1), ('b', 2)]>], value_type = StructType(struct<a: int8, b: int8>)`
- `data = [['a', 'b']], scalar_data = [<pyarrow.ListScalar: ['a', 'b']>], value_type = ListType(list<item: string>)`
- `data = [[1, 2, 3]], scalar_data = [<pyarrow.ListScalar: [1, 2, 3]>], value_type = ListType(list<item: int64>)`

Reply:
Hmm. Sorry for the back and forth here, but then I would maybe go back to the initial skip (with an updated comment). I don't think this is worth the complexity. We know that sets can be iterated multiple times (just with an non-deterministic order), and won't give a problem here, accept being difficult to test with general data. 
(eg, while testing some other options, I also noticed that for the case of null type, for a set it actually created a different array because of de-duplicating the sequence)

Reply:
Agree, will do 👍